_posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md

```diff
@@ -0,0 +1,290 @@
+For the daily operation of the Elasticsearch cluster we implemented a verbose mode in the start script as well as in the `docker-entrypoint` script so that we could monitor each step in the console output.
+
+#### Testing
+
+We versioned the entire source code on [GitHub](https://github.com/). The first file we added was the configuration file for the CircleCI job. The job basically clones the repository and tries to build and run the Docker container. After these described preparation steps several tests check if the Elasticsearch service is reachable from outside of the container and is working as expected. With this setup we could securely develop the `Dockerfile` and the Elasticsearch configuration files against the previously created tests.
+If a pull request was reviewed and merged into the dev branch it will be tested again. If this test run is successful the Docker image will be pushed into our Docker registry and a script merges and pushes the dev code into the master branch.
+
+#### Deployment
+
+Changes in the master branch trigger a Jenkins Job to run, which pulls the latest Docker image from the registry and restarts the container.
+
+### Part 3 - Set up Logstash with Docker and CircleCI
+
+#### Basics
+
+Within our architecture the Logstash container runs as a comprehensive log forwarder, which means it acts as processor, shipper and indexer of the test results. Hence, it's main purpose is to read, transform and feed the test objects as documents into our Elasticsearch cluster. For this purpose the configuration file defines the input and the output of Logstash as well as how the data has to be filtered before dispatching the events to Elasticsearch. The input is given by a file path, which can also contain a regular expression. As described above, we use a JSON log file that is parsed by Logstash. The filter is able to add and remove fields from the JSON object. The output declares where the formatted message should be sent to. In our case this is configurable by environment variables, which will be explained by the following paragraphs.
+
+#### Configuration
+
+In the Logstash configuration, there is the possibility to use `if`-statements and environment variables. In addition to this, we decided to write our own templating engine based on the Jinja2 framework to achieve high flexibility in which variables can be feed into the rendering process. This allows us to have an environment specific configuration for each VM the Docker container is running on. To use this feature we forward some variables into our container. Our `docker-entrypoint` script renders the configuration templates and starts the Logstash agent.
```

Comment:
In the Logstash configuration, there is the possibility to use > In the Logstash configuration it is possible to use